README.rst
==========
STouchTool
==========

.. image:: https://app.travis-ci.com/jtplaarj/STouchTool.svg?branch=master
    :target: https://app.travis-ci.com/jtplaarj/STouchTool
    :alt: Travis
.. image:: https://coveralls.io/repos/github/jtplaarj/STouchTool/badge.svg?branch=master
    :target: https://coveralls.io/github/jtplaarj/STouchTool?branch=master
    :alt: Coveralls
.. image:: https://readthedocs.org/projects/stouchtool/badge/?version=latest
    :target: https://stouchtool.readthedocs.io/en/latest/?badge=latest
    :alt: Documentation Status
.. image:: https://img.shields.io/pypi/v/STouchTool.svg
    :target: https://pypi.org/project/STouchTool/
    :alt: Latest Version
.. image:: http://www.mypy-lang.org/static/mypy_badge.svg
    :target: http://mypy-lang.org/
    :alt: Checked with mypy
.. image:: https://img.shields.io/github/license/jtplaarj/STouchTool?style=flat
    :target: https://mit-license.org/
    :alt: GitHub

Collection of tools to manipulate Touchstone files


Description
===========

This projects aims to collect several utilities to manage Touchstone files. They are mainly wrappers around `scikit-rf <http://scikit-rf.org/>`_ to make them command line friendly.

Commands
========

The list of available commands are:

* ``s_cat``: This command generates an n-port Touchstone file from the appropriate number of two-port files.
* ``s_plot``: This command will plot a Touchstone file into a PDF file.

``s_cat``
---------

This command generates an n-port Touchstone file from the appropriate number of two-port files.

For example, to generate a three-port file from three s2p files::

    s_cat P12_FILE.s2p P13_FILE.s2p P23_FILE.s2p --output output.s3p


The order of the files is important, it must begin with all the combinations of the first port, then the second,...

The number of ports is calculates automatically, if more than 10 ports are used, the explicit number of ports must be provided::

    s_cat *.s2p -n 12 -o test.s12p

The complete list of options is obtained using ``s_cat -h``. The input files to process are mandatory:
    * ``--help, -h``: List of options.
    * ``--numports, -p``: Number of ports, if omitted it will be guessed from number of files.
    * ``--output, -o``: Output file to write result, if none given, it will be the input file with the PDF extension.
    * ``--version``: Package version.
    * ``-v/-vv``: Verbose or very verbose mode.

``s_plot``
----------

This command will plot a Touchstone file into a PDF file.

A simple example is::

    s_plot test.s2p

This will produce a file called ``test.pdf`` plotting the data.

The complete list of options is obtained using ``s_plot -h``. The input file to process are mandatory:
    * ``--help, -h``: List of options.
    * ``--output, -o``: Output file to write result, if none given, it will be the input file with the PDF extension.
    * ``--title, -t``: Title of the plot. If it is not provided, the file name will be used.
    * ``--version``: Package version.
    * ``-v/-vv``: Verbose or very verbose mode.

Installation
============

The easiest way to install this package is to use ``pip``::

    pip install STouchTool


CHANGELOG.rst
=========
Changelog
=========

Version 0.1
===========

- Initial commit


Version 0.1.3
=============

- Corrected workdlow
- Corrected documentation links in packaging
- Corrected names


tests/conftest.py
"""
    Dummy conftest.py for stouchtool.

    If you don't know what this is for, just leave it empty.
    Read more about conftest.py under:
    - https://docs.pytest.org/en/stable/fixture.html
    - https://docs.pytest.org/en/stable/writing_plugins.html
"""

# import pytest


tests/test_s_cat.py
import pytest
import skrf as rf

from stouchtool.s_cat import main, run, s_cat

__author__ = "Jesús Lázaro"
__copyright__ = "Jesús Lázaro"
__license__ = "MIT"


# TODO Test s4p and other properly, with real data
@pytest.mark.parametrize(
    "Inputfile, Outputfile, Golden, ExpectedOutfile, NumPorts",
    [
        (
            [
                "./tests/data/evalboard_in_outp_outn_50ohm_5V_pinm20dBm.s2p",
                "./tests/data/evalboard_in_outn_outp_50ohm_5V_pinm20dBm.s2p",
                "./tests/data/evalboard_outp_outn_in_50ohm_5V_pinm20dBm.s2p",
            ],
            "./tests/data/tmp.s3p",
            "./tests/data/golden.s3p",
            "./tests/data/tmp.s3p",
            None,
        ),
        (
            [
                "./tests/data/evalboard_in_outp_outn_50ohm_5V_pinm20dBm.s2p",
                "./tests/data/evalboard_in_outn_outp_50ohm_5V_pinm20dBm.s2p",
                "./tests/data/evalboard_outp_outn_in_50ohm_5V_pinm20dBm.s2p",
            ],
            "./tests/data/tmp.s3p",
            "./tests/data/golden.s3p",
            "./tests/data/tmp.s3p",
            3,
        ),
        pytest.param(
            [
                "./tests/data/evalboard_in_outp_outn_50ohm_5V_pinm20dBm.s2p",
                "./tests/data/evalboard_in_outn_outp_50ohm_5V_pinm20dBm.s2p",
                "./tests/data/evalboard_outp_outn_in_50ohm_5V_pinm20dBm.s2p",
            ],
            "./tests/data/tmp.s3p",
            "./tests/data/golden.s3p",
            "./tests/data/tmp.s3p",
            4,
            marks=pytest.mark.xfail(raises=ValueError),
        ),
        (
            [
                "./tests/data/evalboard_in_outp_outn_50ohm_5V_pinm20dBm.s2p",
                "./tests/data/evalboard_in_outn_outp_50ohm_5V_pinm20dBm.s2p",
                "./tests/data/evalboard_outp_outn_in_50ohm_5V_pinm20dBm.s2p",
                "./tests/data/evalboard_in_outn_outp_50ohm_5V_pinm20dBm.s2p",
                "./tests/data/evalboard_outp_outn_in_50ohm_5V_pinm20dBm.s2p",
                "./tests/data/evalboard_outp_outn_in_50ohm_5V_pinm20dBm.s2p",
            ],
            "./tests/data/tmp.s4p",
            "./tests/data/golden.s4p",
            "./tests/data/tmp.s4p",
            None,
        ),
        pytest.param(
            [
                "./tests/data/evalboard_in_outp_outn_50ohm_5V_pinm20dBm.s2p",
                "./tests/data/evalboard_outp_outn_in_50ohm_5V_pinm20dBm.s2p",
            ],
            "./tests/data/tmp.s3p",
            "./tests/data/golden.s3p",
            "./tests/data/tmp.s3p",
            None,
            marks=pytest.mark.xfail(raises=ValueError),
        ),
        (
            [
                "./tests/data/evalboard_in_outp_outn_50ohm_5V_pinm20dBm.s2p",
                "./tests/data/evalboard_in_outn_outp_50ohm_5V_pinm20dBm.s2p",
                "./tests/data/evalboard_outp_outn_in_50ohm_5V_pinm20dBm.s2p",
            ],
            None,
            "./tests/data/golden.s3p",
            "./tests/data/evalboard_in_out.s3p",
            None,
        ),
    ],
)
def test_s_cat(
    Inputfile: list, Outputfile: str, Golden: str, ExpectedOutfile: str, NumPorts: int
):
    """Test main function against golden results"""
    CalculatedOutputFile = s_cat(Inputfile, Outputfile, NumPorts)

    assert CalculatedOutputFile == ExpectedOutfile
    GoldenS = rf.Network(Golden)
    ResultS = rf.Network(CalculatedOutputFile)
    assert GoldenS == ResultS


def test_main_args_inputfile(capsys):
    """CLI Tests, file input arguments"""
    # capsys is a pytest fixture that allows asserts agains stdout/stderr
    # https://docs.pytest.org/en/stable/capture.html
    main(
        [
            "./tests/data/evalboard_in_outp_outn_50ohm_5V_pinm20dBm.s2p",
            "./tests/data/evalboard_in_outn_outp_50ohm_5V_pinm20dBm.s2p",
            "./tests/data/evalboard_outp_outn_in_50ohm_5V_pinm20dBm.s2p",
        ]
    )
    captured = capsys.readouterr()
    assert (
        "The cat from files ["
        "'./tests/data/evalboard_in_outp_outn_50ohm_5V_pinm20dBm.s2p', "
        "'./tests/data/evalboard_in_outn_outp_50ohm_5V_pinm20dBm.s2p', "
        "'./tests/data/evalboard_outp_outn_in_50ohm_5V_pinm20dBm.s2p'"
        "] has been stored in "
        "./tests/data/evalboard_in_out.s3p\n" in captured.out
    )


def test_main_no_args(capsys):
    """CLI Tests, no input arguments"""
    with pytest.raises(SystemExit) as pytest_wrapped_e:
        main([])
    captured = capsys.readouterr()
    assert pytest_wrapped_e.type == SystemExit
    assert pytest_wrapped_e.value.code == 2
    assert captured.out == ""
    assert "usage:" in captured.err


def test_main_wrong_args(capsys):
    """CLI Tests, wrong input file number"""
    with pytest.raises(SystemExit) as pytest_wrapped_e:
        main(
            [
                "./tests/data/evalboard_in_outp_outn_50ohm_5V_pinm20dBm.s2p",
                "./tests/data/evalboard_in_outn_outp_50ohm_5V_pinm20dBm.s2p",
            ]
        )
    captured = capsys.readouterr()
    assert pytest_wrapped_e.type == SystemExit
    assert pytest_wrapped_e.value.code == 1
    assert "Wrong number of files" in captured.out


def test_simple_run():
    """Test run entry point"""
    with pytest.raises(SystemExit) as pytest_wrapped_e:
        run()
    assert pytest_wrapped_e.type == SystemExit
    assert pytest_wrapped_e.value.code == 2


tests/test_s_plot.py
import PyPDF2
import pytest

from stouchtool.s_plot import main, run, s_plot

__author__ = "Jesús Lázaro"
__copyright__ = "Jesús Lázaro"
__license__ = "MIT"


@pytest.mark.parametrize(
    "inputfile,outputfile,title,ports, expected_outputfile, expected_title",
    [
        (
            "./tests/data/limiter_pin_0dBm.s2p",
            None,
            None,
            2,
            "./tests/data/limiter_pin_0dBm.pdf",
            "limiter_pin_0dBm",
        ),
        (
            "./tests/data/evalboard.s3p",
            "./tests/data/test.pdf",
            None,
            3,
            "./tests/data/test.pdf",
            "evalboard",
        ),
        (
            "./tests/data/limiter_pin_0dBm.s2p",
            "./tests/data/test.pdf",
            "Nice title",
            2,
            "./tests/data/test.pdf",
            "Nice title",
        ),
    ],
)
def test_s_plot_ok(
    inputfile: str,
    outputfile: str,
    title: str,
    ports: int,
    expected_outputfile: str,
    expected_title: str,
):
    """Test the API with correct values - no exception should occur"""

    (inputfile_result, outputfile_result, ports_result) = s_plot(
        inputfile, outputfile, title
    )
    assert (inputfile_result, outputfile_result, ports_result) == (
        inputfile,
        expected_outputfile,
        ports,
    )
    pdfFileObject = open(expected_outputfile, "rb")
    pdfReader = PyPDF2.PdfFileReader(pdfFileObject)
    """ PDF should have a single page """
    assert pdfReader.numPages == 1
    """ Check for title in page"""
    pageObject = pdfReader.getPage(0)
    assert pageObject.extractText().find(expected_title) > -1


def test_s_plot_except():
    """Test the API with non existing file to check for error"""
    with pytest.raises(SystemExit):
        # non existing file
        assert s_plot("kk.s3p", "./tests/data/test.pdf", None) == (
            "./tests/data/evalboard.s2p",
            "./tests/data/test.pdf",
            2,
        )


def test_main(capsys):
    """CLI Tests"""
    # capsys is a pytest fixture that allows asserts agains stdout/stderr
    # https://docs.pytest.org/en/stable/capture.html
    main(["./tests/data/evalboard.s3p"])
    captured = capsys.readouterr()
    assert (
        "The plot from file ./tests/data/evalboard.s3p has 3 ports and has "
        "been ploted in ./tests/data/evalboard.pdf\n" in captured.out
    )


def test_main_no_args(capsys):
    """CLI Tests, no input arguments"""
    with pytest.raises(SystemExit) as pytest_wrapped_e:
        main([])
    captured = capsys.readouterr()
    assert pytest_wrapped_e.type == SystemExit
    assert pytest_wrapped_e.value.code == 2
    assert captured.out == ""
    assert "usage:" in captured.err


def test_simple_run():
    """Test run entry point"""
    with pytest.raises(SystemExit) as pytest_wrapped_e:
        run()
    assert pytest_wrapped_e.type == SystemExit
    assert pytest_wrapped_e.value.code == 2


src/stouchtool/s_plot.py
"""
This script aims to pretty plot s params obtained from VSA or Python scikit-rf


References:
    - https://scikit-rf.readthedocs.io
"""

import argparse
import logging
import os
import sys
from typing import List, Tuple

import matplotlib.pyplot as plt
import skrf as rf
from matplotlib import ticker

from stouchtool import __version__

__author__ = "Jesús Lázaro"
__copyright__ = "Jesús Lázaro"
__license__ = "MIT"

_logger = logging.getLogger(__name__)


# ---- Python API ----
# The functions defined in this section can be imported by users in their
# Python scripts/interactive interpreter, e.g. via
# `from STouchTool.skeleton import fib`,
# when using this Python module as a library.


def s_plot(input: str, output: str, title: str) -> Tuple[str, str, int]:
    """Generate a plot in pdf with the provided touchstone data

    Args:
        input (str): input file name
        output (str): output file name, if none, it will be derived from input
        title (str): title of the plot, if none, it will derived from input

    Returns:
        Tuple[str, str, int]: input file name, output file name, number of ports
    """

    _logger.info("s_plot: The input file is:{}".format(input))

    # output file may be none, generate correct one
    if output is None:
        _logger.info(
            "s_plot: The output file is not given so a new one will be created"
        )
        output = os.path.splitext(input)[0] + ".pdf"
    _logger.info("s_plot: The output file is:{}".format(output))
    if title is None:
        _logger.info("s_plot: The title is not given so a new one will be created")
        title = os.path.splitext(os.path.basename(input))[0]

    _logger.info("s_plot: The title is:{}".format(title))

    try:
        slot = rf.Network(input)
    except Exception as e:
        _logger.debug("s_plot: Exception {} when opening file: {}".format(e, input))
        sys.exit(1)

    fig, ax = plt.subplots()
    slot.plot_s_db(ax=ax)
    ax.yaxis.set_minor_locator(ticker.MultipleLocator(base=5.0))
    ax.xaxis.set_minor_locator(ticker.MultipleLocator(base=100e6))
    ax.xaxis.set_major_formatter(
        ticker.FuncFormatter(lambda x, pos: "{:.0f}".format(x / 1e6))
    )

    ax.grid(which="major", color="#CCCCCC", linestyle="--")
    ax.grid(which="minor", color="#CCCCCC", linestyle=":")

    ax.set_xlabel("Frequency (MHz)")
    ax.set_title(title)

    # Change legend names since the default are very long names with the filename in it
    legend_list = []
    for output_port in range(slot.number_of_ports):
        for input_port in range(slot.number_of_ports):
            legend_list.append("S{}{}".format(output_port + 1, input_port + 1))

    ax.legend(legend_list)

    plt.savefig(output, format="pdf", bbox_inches="tight")

    return (input, output, slot.number_of_ports)


# ---- CLI ----
# The functions defined in this section are wrappers around the main Python
# API allowing them to be called directly from the terminal as a CLI
# executable/script.


def parse_args(args: List[str]) -> argparse.Namespace:
    """Parse command line parameters

    Args:
        args (List[str]): command line parameters as list of strings

    Returns:
        :obj:`argparse.Namespace`: command line parameters namespace
    """

    parser = argparse.ArgumentParser(description="Plot S params")
    parser.add_argument(
        "--version",
        action="version",
        version="STouchTool {ver}".format(ver=__version__),
    )
    parser.add_argument(
        dest="input",
        help="Input file with touchstone params",
        type=str,
        metavar="INPUT FILE",
    )
    parser.add_argument(
        "-o",
        "--output",
        dest="output",
        help="Output file to write result, if none given, \
            it will be the input file with the PDF extension",
        type=str,
        metavar="OUTPUT FILE",
    )
    parser.add_argument(
        "-t",
        "--title",
        dest="title",
        help="Title of the plot",
        type=str,
        metavar="TITLE",
    )
    parser.add_argument(
        "-v",
        "--verbose",
        dest="loglevel",
        help="set loglevel to INFO",
        action="store_const",
        const=logging.INFO,
    )
    parser.add_argument(
        "-vv",
        "--very-verbose",
        dest="loglevel",
        help="set loglevel to DEBUG",
        action="store_const",
        const=logging.DEBUG,
    )
    return parser.parse_args(args)


def setup_logging(loglevel: int):
    """setup logging

    Args:
        loglevel (int): minimum loglevel for emitting messages
    """


def main(arguments: List[str]):
    """Wrapper allowing :func:`s_plot` to be called as CLI

    Args:
        arguments (List[str]): command line parameters as list of strings
    """

    args = parse_args(arguments)
    setup_logging(args.loglevel)
    _logger.debug("Starting plotting...")
    inputfilename, outputfilename, numberofports = s_plot(
        args.input, args.output, args.title
    )
    print(
        "The plot from file {} has {} ports and has been ploted in {}".format(
            inputfilename, numberofports, outputfilename
        )
    )
    _logger.info("s_plot: Script ends here")


def run():
    """Calls :func:`main` passing the CLI arguments extracted from :obj:`sys.argv`

    This function can be used as entry point to create console scripts with setuptools.
    """
    main(sys.argv[1:])


if __name__ == "__main__":
    # ^  This is a guard statement that will prevent the following code from
    #    being executed in the case someone imports this file instead of
    #    executing it as a script.
    #    https://docs.python.org/3/library/__main__.html

    # After installing your project with pip, users can also run your Python
    # modules as scripts via the ``-m`` flag, as defined in PEP 338::
    #
    #     python -m STouchTool.skeleton 42
    #
    run()


src/stouchtool/s_cat.py
# Copyright (c) 2021 Jesús Lázaro
#
# This software is released under the MIT License.
# https://opensource.org/licenses/MIT

import argparse
import logging
import sys
from difflib import SequenceMatcher
from typing import List

import skrf as rf
from scipy.special import comb

from stouchtool import __version__

__author__ = "Jesús Lázaro"
__copyright__ = "Jesús Lázaro"
__license__ = "MIT"

_logger = logging.getLogger(__name__)


def s_cat(inputfiles: List[str], outputfile: str, NumPort: int) -> str:
    """Concatenate 2 port s files into an n port s file

    Args:
        inputfiles (List[str]): List of files
        outputfile (str): Name of output file - optional
        NumPort (int): Number of ports - optional

    Raises:
        ValueError: In provided number of ports and files do not match

    Returns:
        str: final output file name
    """

    NumFiles = len(inputfiles)
    NumPortFound = False
    if NumPort is None:
        for NumPort in range(1, 10):
            if comb(NumPort, 2) == NumFiles:
                NumPortFound = True
                break
    else:
        if comb(NumPort, 2) == NumFiles:
            NumPortFound = True

    if not NumPortFound:
        _logger.debug("Wrong number of files: {}".format(NumFiles))
        raise ValueError("Wrong number of files: {}".format(NumFiles))

    PortList = list()
    RFNetworks = list()
    for OutputPort in range(1, NumPort + 1):
        for InputPort in range(OutputPort + 1, NumPort + 1):
            PortList.append("p" + str(OutputPort) + str(InputPort))
    _logger.debug(
        "Number of files is {} and number of ports is {}".format(NumFiles, NumPort)
    )
    for (inputfile, port) in zip(inputfiles, PortList):
        _logger.debug("File {} is {}".format(inputfile, port))
        tmpNetwork = rf.Network(inputfile)
        tmpNetwork.name = port
        RFNetworks.append(tmpNetwork)

    if outputfile is None:
        _logger.debug("The output file is not given so a new one will be created")
        match = SequenceMatcher(None, inputfiles[0], inputfiles[1]).find_longest_match(
            0, len(inputfiles[0]), 0, len(inputfiles[1])
        )
        outputfile = (
            inputfiles[0][match.a : match.a + match.size] + ".s" + str(NumPort) + "p"
        )

    _logger.debug("Combining: {}".format(RFNetworks))
    combined = rf.network.n_twoports_2_nport(RFNetworks, nports=NumPort)
    combined.write_touchstone(outputfile)
    return outputfile


def parse_args(args: List[str]) -> argparse.Namespace:
    """Parse command line parameters

    Args:
        args (List[str]): command line parameters as list of strings

    Returns:
        :obj:`argparse.Namespace`: command line parameters namespace
    """

    parser = argparse.ArgumentParser(description="Concatenate S params")
    parser.add_argument(
        "--version",
        action="version",
        version="RFTools {ver}".format(ver=__version__),
    )
    parser.add_argument(
        dest="inputfiles",
        help="Input file with touchstone params",
        type=str,
        nargs="+",
        metavar="P12_FILE.s2p P13_FILE.s2p P23_FILE.s2p",
    )
    parser.add_argument(
        "-p",
        "--numports",
        dest="numports",
        help="Number of ports, if ommited it will be gussed from number of files",
        type=int,
        metavar="NUM_PORTS",
    )
    parser.add_argument(
        "-o",
        "--output",
        dest="output",
        help="Output file to write result, if none given, \
            it will be the input file with the PDF extension",
        type=str,
        metavar="OUTPUT_FILE",
    )
    parser.add_argument(
        "-v",
        "--verbose",
        dest="loglevel",
        help="set loglevel to INFO",
        action="store_const",
        const=logging.INFO,
    )
    parser.add_argument(
        "-vv",
        "--very-verbose",
        dest="loglevel",
        help="set loglevel to DEBUG",
        action="store_const",
        const=logging.DEBUG,
    )
    return parser.parse_args(args)


def setup_logging(loglevel: int):
    """setup logging

    Args:
        loglevel (int): minimum loglevel for emitting messages
    """

    logformat = "[%(asctime)s] %(levelname)s:%(name)s:%(message)s"
    logging.basicConfig(
        level=loglevel, stream=sys.stdout, format=logformat, datefmt="%Y-%m-%d %H:%M:%S"
    )


def main(arguments: List[str]):
    """Wrapper allowing :func:`plot_s_cat` to be called as CLI

    Args:
        args (List[str]): command line parameters as list of strings
    """

    args = parse_args(arguments)
    setup_logging(args.loglevel)
    _logger.debug("Starting plotting...")

    try:
        outputfilename = s_cat(args.inputfiles, args.output, args.numports)
    except ValueError as e:
        print(e)
        sys.exit(1)
    print(
        "The cat from files {} has been stored in {}".format(
            args.inputfiles, outputfilename
        )
    )
    _logger.info("plot_ad_data: Script ends here")


def run():
    """Calls :func:`main` passing the CLI arguments extracted from :obj:`sys.argv`

    This function can be used as entry point to create console scripts with setuptools.
    """
    main(sys.argv[1:])


if __name__ == "__main__":
    # ^  This is a guard statement that will prevent the following code from
    #    being executed in the case someone imports this file instead of
    #    executing it as a script.
    #    https://docs.python.org/3/library/__main__.html

    # After installing your project with pip, users can also run your Python
    # modules as scripts via the ``-m`` flag, as defined in PEP 338::
    #
    #     python -m rftools.skeleton 42
    #
    run()
